numpy/f2py/__init__.py

```diff
@@ -48,25 +56,49 @@ def compile(source,
         .. versionadded:: 1.11.0
 
     """
-    from numpy.distutils.exec_command import exec_command
     import tempfile
+    import shlex
+
     if source_fn is None:
-        f = tempfile.NamedTemporaryFile(suffix=extension)
+        f, fname = tempfile.mkstemp(suffix=extension)
+        # f is a file descriptor so need to close it
+        # carefully -- not with .close() directly
+        os.close(f)
     else:
-        f = open(source_fn, 'w')
+        fname = source_fn
 
     try:
-        f.write(source)
-        f.flush()
-
-        args = ' -c -m {} {} {}'.format(modulename, f.name, extra_args)
-        c = '{} -c "import numpy.f2py as f2py2e;f2py2e.main()" {}'
-        c = c.format(sys.executable, args)
-        status, output = exec_command(c)
-        if verbose:
-            print(output)
+        with open(fname, 'w') as f:
+            f.write(str(source))
```

Comment:
I have to coerce with `str` here when I work on my mac to get the tests added in this PR to pass

Reply:
removing the flush line (with no other changes) causes `test_f2py_init_compile()` to fail on Ubuntu for all 3 parametrizations with: `error: f2py target file '/tmp/tmpep0xud0c/src.linux-x86_64-3.6/_test_ext_module_5411module.c' not generated`

I restored the flush line and failures were again suppressed in full test suite.